packages/core/src/directory.ts

```diff
@@ -0,0 +1,91 @@
+import { Org, OrgID, OrgMember, OrgMemberStatus } from "./org";
+import { uuid } from "./util";
+import { Storage } from "./storage";
+
+export interface DirectoryUser {
+    email: string;
+    name: string;
+    active: boolean;
+}
+
+export interface DirectoryGroup {
+    name: string;
+    members: DirectoryUser[];
+}
+
+export interface DirectorySubscriber {
+    userCreated(user: DirectoryUser, orgId?: OrgID): Promise<void>;
+    userUpdated(user: DirectoryUser, orgId?: OrgID): Promise<void>;
+    userDeleted(user: DirectoryUser, orgId?: OrgID): Promise<void>;
+
+    groupCreated(group: DirectoryGroup, orgId: OrgID): Promise<void>;
+    groupUpdated(group: DirectoryGroup, orgId: OrgID): Promise<void>;
+    groupDeleted(group: DirectoryGroup, orgId: OrgID): Promise<void>;
+}
+
+export interface DirectoryProvider {
+    subscribe(sub: DirectorySubscriber): void;
+}
+
+export class DirectorySync implements DirectorySubscriber {
+    constructor(public readonly storage: Storage, providers: DirectoryProvider[] = []) {
+        for (const provider of providers) {
+            provider.subscribe(this);
+        }
+    }
+
+    async userCreated(user: DirectoryUser, orgId: string) {
+        const org = (orgId && (await this.storage.get(Org, orgId))) || null;
+        if (org && org.directory.syncProvider === "scim" && org.directory.syncMembers) {
+            org.members.push(
+                new OrgMember({
+                    email: user.email,
+                    status: OrgMemberStatus.Provisioned,
+                    updated: new Date(),
+                })
+            );
+            org.revision = await uuid();
+            org.updated = new Date();
+
+            await this.storage.save(org);
+
+            // TODO: Remove this
+            console.log("---- core/src/directory. userCreated");
+            console.log(org.members);
```

Comment:
@MaKleSoft now this seems to be saving, but when I load up the org again, it doesn't have the new org member (I also see no invites or anything that should've been auto-created), and no errors. I haven't had a chance to debug that, though.

Reply:
I think I know what the problem is. Let me take a closer look.

Reply:
Ok, so updating the `revision` property on the `Org` isn't enough, you actually have to update the `orgs` property for all member accounts as well. Actually that's what [`updateMetaData`](https://github.com/padloc/padloc/blob/master/packages/core/src/server.ts#L1352) in the `server.ts` module is for. In fact, I'm thinking it may be better to pass the `Server` instance to `DirectorySync` instead of just the storage. Then you can do:

```ts
this.server.updateMetaData(org);
this.server.storage.save(org);
```

Reply:
> Ok, so updating the revision property on the Org isn't enough, you actually have to update the orgs property for all member accounts as well.

To understand why this is necessary, take a look [here](https://github.com/padloc/padloc/blob/acbc931b510df53dfaaf0b96fb27cad8463e99f1/packages/core/src/app.ts#L1478) and [here](https://github.com/padloc/padloc/blob/acbc931b510df53dfaaf0b96fb27cad8463e99f1/packages/core/src/app.ts#L1489).

Reply:
I'm afraid this wasn't enough. I've investigated this deeper and I believe the problem comes from [here](https://github.com/padloc/padloc/blob/feature/scim-v1/packages/core/src/server.ts#L1999-L2001), because there is no `account.orgs` to save the data into and it doesn't get saved?

Reply:
@BrunoBernardino Can you give me an example SCIM payload so I can test this locally?

Reply:
Sure, check the initial comment, it's got one, though you'll need to change the URL to the one you get in your org's settings after enabling SCIM.

If it helps, I boot the app with `PL_PROVISIONING_BACKEND=directory PL_DIRECTORY_PROVIDERS=scim PL_DIRECTORY_SCIM_PORT=5000 npm run dev`